x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/ApplicationPrivilege.java

```diff
@@ -29,8 +29,16 @@
  */
 public final class ApplicationPrivilege extends Privilege {
 
-    private static final Pattern VALID_APPLICATION = Pattern.compile("^[a-z][A-Za-z0-9_-]{2,}$");
-    private static final Pattern VALID_APPLICATION_OR_WILDCARD = Pattern.compile("^[a-z*][A-Za-z0-9_*-]*");
+    /**
+     * Application name validation. App names must
+     * - Start with a lower case letter
+     * - Followed by 2 or more alpha-numeric ASCII characters
+     * - Optionally followed by:
+     *   - A single "-" or "_"
+     *   - Then any number of printable, visible ASCII characters (letter, numbers, symbols) other than '*'
+      */
+    private static final Pattern VALID_APPLICATION = Pattern.compile("^[a-z][A-Za-z0-9]{2,}([_-][\\p{Graph}&&[^*]]*)?$");
+    private static final Pattern VALID_APPLICATION_OR_WILDCARD = Pattern.compile("^[a-z*][A-Za-z0-9*]*([_-]\\p{Graph}*)?$");
```

Comment:
The \\p{Graph} thing is probably too broad for indices names https://github.com/elastic/elasticsearch/blob/49b977ba7ca482d518f3001221a117baa3ffe2ae/server/src/main/java/org/elasticsearch/common/Strings.java#L376 . I would leave it as it is though, treading on regex is long haul.

`VALID_APPLICATION_OR_WILDCARD` is again too broad, it accepts multiple * intertwined with text, so more like a regex than an wildcard.
This one I would change. I would use the same regex as `VALID_APPLICATION` but ending with `[*]?$` instead of `$`. Also a test for this would be nice.

Reply:
Are you suggesting that we only support wildcards that end in `*`?
That wasn't my intent, but it solves all the use cases for which wildcards are required, so it sounds like a reasonable approach, although I do need to work out an appropriate regex for it.

I do wonder whether trying to do all this with a regex is becoming a little futile...

Reply:
> Also, a test for this would be nice.

What specifically did you want a test for? There are tests for wildcards
- https://github.com/elastic/elasticsearch/pull/31752/files#diff-965f80992d9820eff5becc043e4cb96aR42
- https://github.com/elastic/elasticsearch/pull/31752/files#diff-965f80992d9820eff5becc043e4cb96aR50
- https://github.com/elastic/elasticsearch/pull/31752/files#diff-965f80992d9820eff5becc043e4cb96aR59

Reply:
Yes, I meant the `*` to only be allowed at the end, and the test will check that `*` is not allowed anywhere else. It was a reflex, I haven't considered the implication of it, though...